### PR TITLE
Shashank/signature shim

### PR DESCRIPTION
**Pull request by sudo-shashank**
## Summary of changes

Changes introduced in this pull request:
- Added `Signature` and `SignatureType` type in forest_shim that can convert to FVM2 and FVM3.

## Reference issue to close (if applicable)
Part of https://github.com/ChainSafe/forest/issues/2438

## Change checklist

- [X] I have performed a self-review of my own code,
- [X] I have made corresponding changes to the documentation,
- [X] I have added tests that prove my fix is effective or that my feature works (if possible),
- [X] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

**Comment by lemmih** [comment]
@sudo-shashank Do you want to move forward with this PR?

**Comment by sudo-shashank** [comment]
> @sudo-shashank Do you want to move forward with this PR?

Yes signature verify required address shim so I was waiting for it to be completed

**Comment by lemmih** [comment]
> > @sudo-shashank Do you want to move forward with this PR?
> 
> Yes signature verify required address shim so I was waiting for it to be completed

Okay. The address shim is done but you're gonna have a lot of conflicts.

**Comment by lemmih** [comment]
@sudo-shashank It may be that we don't need to shim the `Signature` type. Let's wait and see.

**Comment by sudo-shashank** [comment]
> @sudo-shashank It may be that we don't need to shim the `Signature` type. Let's wait and see.

all changes are made for Signature and SignatureType, if this is required we can review and merge
keeping it as draft PR for now as we are not sure

File: utils/forest_shim/src/crypto.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Copy, Serialize, Deserialize)]
+#[repr(transparent)]
+#[serde(transparent)]
+pub struct SignatureType(pub SignatureType_v3);
+
+impl SignatureType {
+    pub const BLS: Self = Self(SignatureType_v3::BLS);
+    pub const SECP256K1: Self = Self(SignatureType_v3::Secp256k1);
```
Comment (lemmih) [inline utils/forest_shim/src/crypto.rs:100]:
```suggestion
    #[allow(non_upper_case_globals)]
    pub const Secp256k1: Self = Self(SignatureType_v3::Secp256k1);
```

File: blockchain/chain/src/store/chain_store.rs [hunk 2]
```diff
@@ -938,7 +940,7 @@ pub fn persist_block_messages<DB: Blockstore>(
 
     let mut bls_sigs = Vec::new();
     for msg in messages {
-        if msg.signature().signature_type() == SignatureType::BLS {
+        if msg.signature().signature_type() == SignatureType::BLS.into() {
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:943]: Change `SignedMessage::signature()` to return a shimmed signature.

**Review by lemmih — CHANGES_REQUESTED** [review]

**Review by LesnyRumcajs — APPROVED** [review]

File: key_management/src/wallet_helpers.rs [hunk 3]
```diff
@@ -52,7 +54,7 @@ pub fn sign(sig_type: SignatureType, private_key: &[u8], msg: &[u8]) -> Result<S
             // this returns a signature from bls-signatures, so we need to convert this to a
             // crypto signature
             let sig = priv_key.sign(msg);
-            let crypto_sig = Signature::new_bls(sig.as_bytes());
+            let crypto_sig = Signature::new(SignatureType::BLS, sig.as_bytes());
```
Comment (lemmih) [inline key_management/src/wallet_helpers.rs:57]: Can't you use `new_bls` here?

File: utils/json/src/signature.rs [hunk 4]
```diff
@@ -30,7 +30,7 @@ pub mod json {
         S: Serializer,
     {
         JsonHelper {
-            sig_type: m.sig_type,
+            sig_type: m.sig_type.into(),
```
Comment (lemmih) [inline utils/json/src/signature.rs:33]:
```suggestion
            sig_type: m.signature_type(),
```

File: vm/message/tests/builder_test.rs [hunk 5]
```diff
@@ -25,5 +25,5 @@ fn generate_signed_message() {
 
     // Assert message and signature are expected
     assert_eq!(signed_msg.message(), &msg);
-    assert_eq!(signed_msg.signature(), &Signature::new_secp256k1(dummy_sig));
+    assert_eq!(*signed_msg.signature(), Signature::new_secp256k1(dummy_sig));
```
Comment (lemmih) [inline vm/message/tests/builder_test.rs:28]: Is this necessary?

File: key_management/src/wallet_helpers.rs (hunk 3 above)
Comment (sudo-shashank) [reply key_management/src/wallet_helpers.rs]: fixed

File: key_management/src/wallet_helpers.rs [hunk 6]
```diff
@@ -64,7 +66,7 @@ pub fn sign(sig_type: SignatureType, private_key: &[u8], msg: &[u8]) -> Result<S
             let mut new_bytes = [0; 65];
             new_bytes[..64].copy_from_slice(&sig.serialize());
             new_bytes[64] = recovery_id.serialize();
-            let crypto_sig = Signature::new_secp256k1(new_bytes.to_vec());
+            let crypto_sig = Signature::new(SignatureType::Secp256k1, new_bytes.to_vec());
```
[the remark this answers is not in the document]
Comment (sudo-shashank) [reply key_management/src/wallet_helpers.rs]: fixed

File: utils/json/src/signature.rs (hunk 4 above)
Comment (sudo-shashank) [reply utils/json/src/signature.rs]: fixed

File: vm/message/tests/builder_test.rs (hunk 5 above)
Comment (sudo-shashank) [reply vm/message/tests/builder_test.rs]: no fixed

File: blockchain/chain/src/store/chain_store.rs [hunk 7]
```diff
@@ -941,7 +943,7 @@ pub fn persist_block_messages<DB: Blockstore>(
...
```
Comment (lemmih) [inline blockchain/chain/src/store/chain_store.rs:946]: You can add something like:
```rust
pub fn signature_type(&self) -> SignatureType {
    self.0.signature_type().into()
}
```
to `Signature`.

File: utils/forest_shim/src/crypto.rs [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            bytes,
+        })
+    }
+
+    /// Creates a BLS Signature given the raw bytes.
+    pub fn new_bls(bytes: Vec<u8>) -> Self {
+        Signature(Signature_v3 {
+            sig_type: SignatureType_v3::BLS,
+            bytes,
+        })
```
Comment (lemmih) [inline utils/forest_shim/src/crypto.rs:31]:
```suggestion
        Signature(Signature_v3::new_bls(bytes))
```

File: utils/forest_shim/src/crypto.rs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    /// Creates a SECP Signature given the raw bytes.
+    pub fn new_secp256k1(bytes: Vec<u8>) -> Self {
+        Signature(Signature_v3 {
+            sig_type: SignatureType_v3::Secp256k1,
+            bytes,
+        })
```
Comment (lemmih) [inline utils/forest_shim/src/crypto.rs:39]:
```suggestion
        Signature(Signature_v3::new_secp256k1(bytes))
```

**Review by lemmih — CHANGES_REQUESTED** [review]

File: blockchain/chain/src/store/chain_store.rs (hunk 7 above)
Comment (sudo-shashank) [reply blockchain/chain/src/store/chain_store.rs:946]: fixed

File: utils/forest_shim/src/crypto.rs (hunk 8 above)
Comment (sudo-shashank) [reply utils/forest_shim/src/crypto.rs]: done

File: utils/forest_shim/src/crypto.rs (hunk 9 above)
Comment (sudo-shashank) [reply utils/forest_shim/src/crypto.rs]: done

File: blockchain/chain/src/store/chain_store.rs (hunk 7 above)
Comment (lemmih) [reply blockchain/chain/src/store/chain_store.rs]: Still need to propagate that change. See the lint errors.

File: blockchain/message_pool/src/msgpool/msg_pool.rs [hunk 10]
```diff
@@ -590,7 +593,7 @@ pub(crate) fn add_helper<T>(
 where
     T: Provider,
 {
-    if msg.signature().signature_type() == SignatureType::BLS {
+    if msg.signature().signature_type() == SignatureType::BLS.into() {
```
Comment (lemmih) [inline blockchain/message_pool/src/msgpool/msg_pool.rs:596]:
```suggestion
    if msg.signature().signature_type() == SignatureType::BLS {
```

File: utils/json/src/signature.rs [hunk 11]
```diff
@@ -30,7 +30,7 @@ pub mod json {
         S: Serializer,
     {
         JsonHelper {
-            sig_type: m.sig_type,
+            sig_type: m.signature_type().into(),
```
Comment (lemmih) [inline utils/json/src/signature.rs:33]:
```suggestion
            sig_type: m.signature_type(),
```

File: vm/message/src/signed_message.rs [hunk 12]
```diff
@@ -52,12 +54,12 @@ impl SignedMessage {
 
     /// Checks if the signed message is a BLS message.
     pub fn is_bls(&self) -> bool {
-        self.signature.signature_type() == SignatureType::BLS
+        self.signature.signature_type() == SignatureType::BLS.into()
```
Comment (lemmih) [inline vm/message/src/signed_message.rs:57]:
```suggestion
        self.signature.signature_type() == SignatureType::BLS
```

File: vm/message/src/signed_message.rs [hunk 13]
```diff
@@ -52,12 +54,12 @@ impl SignedMessage {
...
     }
 
     /// Checks if the signed message is a SECP message.
     pub fn is_secp256k1(&self) -> bool {
-        self.signature.signature_type() == SignatureType::Secp256k1
+        self.signature.signature_type() == SignatureType::Secp256k1.into()
```
Comment (lemmih) [inline vm/message/src/signed_message.rs:62]:
```suggestion
        self.signature.signature_type() == SignatureType::Secp256k1
```

**Review by lemmih — APPROVED** [review]

**Pull request merged**